tutorials/03_moduly.md

```diff
@@ -0,0 +1,346 @@
+        'Programming Language :: Python :: 3.5',
+        'Topic :: Software Development :: Libraries',
+        ]
+)
+```
+
+Všimněte si několika věcí. V první řadě v `long_description` vidíte, že jsme
+pořád ještě v Pythonu a můžeme si ušetřit duplikaci nějakých informací pomocí
+malého kousku kódu. Dalším zajímavým argumentem je `classifiers`. Jsou to
+v podstatě takové tagy nebo strukturované informace o balíčku.
+Nevymýšlíme je sami, ale hledáme je v
+[seznamu](https://pypi.python.org/pypi?%3Aaction=list_classifiers).
+Tyto informace jsou později vidět na [PyPI](https://pypi.python.org/pypi).
+
+Více souborů s Python kódem
+---------------------------
+
+Doteď jsme vytvářeli balíček jen z jednoho zdrojového souboru `isholiday.py`.
+Co ale dělat, pokud je náš projekt větší a obsahuje souborů více?
+Teoreticky je možné je přidat všechny do `py_modules`, ale není to dobrý nápad.
+
+V takovém případě uděláme modul ve formě složky. V našem případě soubor
+`isholiday.py` zatím přesuneme do `isholiday/__init__.py`:
+
+```bash
+(env)$ tree
+.
+├── isholiday
+│   └── __init__.py
+├── LICENSE
+├── MANIFEST.in
+├── README
+└── setup.py
+
+1 directory, 5 files
+```
+
+Musíme ještě mírně upravit `setup.py`:
+
+```patch
+diff --git a/setup.py b/setup.py
+index 3a69792..6b453ab 100644
+--- a/setup.py
++++ b/setup.py
+@@ -11,7 +11,7 @@ setup(
+     keywords='holiday,dates',
+     license='Public Domain',
+     url='https://gist.github.com/oskar456/e91ef3ff77476b0dbc4ac19875d0555e',
+-    py_modules=['isholiday'],
++    packages=['isholiday'],
+     classifiers=[
+         'Intended Audience :: Developers',
+         'License :: Public Domain',
+```
+
+Případně, což je ještě lepší, můžeme použít `find_packages()`:
+
+```python
+from setuptools import setup, find_packages
+
+setup(
+    ...
+    packages=find_packages(),
+    ...
+)
+```
+
+Momentálně máme všechen kód přímo v `__init__.py`, což sice funguje,
+ale ideální to není. Dobré je mít kód v samostatných souborech a v `__init__.py`
+pouze importovat veřejné rozhraní, například takto:
+
+```python
+from .holidays import getholidays, isholiday
+
+__all__ = ['getholidays', 'isholiday']
+
+```
+
+Do `__init__.py` ideálně nepatří žádný kód kromě tohoto.
+
+Programy pro příkazovou řádku
+-----------------------------
+
+Pokud chcete, aby váš modul umožňoval spouštění z příkazové řádky, měli byste
+použít [entrypoints]:
+
+[entrypoints]: https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins
+
+```python
+setup(
+    entry_points={
+        'console_scripts': [
+            'executable_name = isholiday.__main__:main'
+        ]
+    },
+)
+```
+
+`isholiday.__main__:main` je cesta k funkci ve tvaru `modul:funkce`, funkce může
+být v modulu definovaná nebo importovaná.
+
+Všimněte si, že je zde použito `__main__`, což předpokládá existenci souboru
+`__main__.py` v modulu. Tento soubor by měl vykonávat funkci skriptu.
```

Comment:
"vykonávat funkci skriptu" znamená, že by měla mít na konci:

```
    if __name__ == '__main__':
        main()
```

(FWIW, já často `console_scripts` nepoužívám, a místo toho se to spouští přes `python -m module_name`. Pak tam nemusí být funkce `main()`, ani ten `if` – `__main__` se odjinud neimportuje. Je dobré to studentům říct, nebo je spíš nemást?)